Add runtime handler seccomp profile

gavinkflam opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->

<!--
/kind api-change
/kind bug
/kind ci
/kind cleanup
/kind dependency-change
/kind deprecation
/kind design
/kind documentation
/kind failing-test
/kind feature
/kind flake
/kind other
-->

/kind feature

#### What this PR does / why we need it:

Allow configuration of a default seccomp profile per runtime.

#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->

Fixes #5715

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Allow configuration of a default seccomp profile per runtime.
```

gavinkflam commented:
/cc saschagrunert haircommander

Please take a look if you get a chance.

saschagrunert commented:
/ok-to-test

saschagrunert reviewed (commented):
Thanks! I just have a few suggestions, otherwise LGTM

gavinkflam commented on test/seccomp_runtime_handler.bats:
```diff
@@ -0,0 +1,49 @@
+	crictl inspect "$CTR" | jq -e .info.runtimeSpec.linux.seccomp | grep -q $TEST_SYSCALL
+}
+
+@test "seccomp runtime handler profile file is invalid" {
```
Added two more test cases.

gavinkflam commented on pkg/config/template.go:
```diff
@@ -1300,6 +1301,8 @@ const templateStringCrioRuntimeRuntimesRuntimeHandler = `# The "crio.runtime.run
 #   when a machine crash happens.
 # - default_annotations (optional, map): Default annotations if not overridden by the pod spec.
 # - stream_websockets (optional, bool): Enable the WebSocket protocol for container exec, attach and port forward.
+# - seccomp_profile (optional, string): The path of the seccomp.json profile which is used as the default seccomp profile
```
Done.

gavinkflam commented on pkg/config/config.go:
```diff
@@ -1999,6 +2009,19 @@ func (r *RuntimeHandler) validateRuntimeExecCPUAffinity() error {
 }
 
+// ValidateRuntimeSeccompProfile checks if the RuntimeHandler seccomp profile exists.
+func (r *RuntimeHandler) ValidateRuntimeSeccompProfile(name string) error {
```
Added absolute path validation and a unit test.

bitoku commented on pkg/config/config.go:
```diff
@@ -288,13 +288,18 @@ type RuntimeHandler struct {
+
+	// SeccompProfile is the seccomp.json profile path which is used as the
+	// default for the runtime.
+	SeccompProfile string `toml:"seccomp_profile,omitempty"`
```
It looks ok to me, but want it to say that this supersedes SeccompProfile in RuntimeConfig, and fall back to it if it's empty.
I'm wondering if we want the one for the privileged pod, which is in RuntimeConfig.

gavinkflam replied:
I’ve revised multiple comments to better explain the precedence.

`PrivilegedSeccompProfile` - I'd be glad to submit a follow-up pull request if that is what we want.

gavinkflam commented:
/retest

gavinkflam commented on pkg/config/config.go:
```diff
@@ -1999,6 +2012,23 @@ func (r *RuntimeHandler) validateRuntimeExecCPUAffinity() error {
+	}
+
+	return nil
+}
```
This is done.

gavinkflam commented:
I have aligned the implementation with `RuntimeConfig.SeccompProfile`. This ensures consistent and predictable behavior as the `RuntimeConfig` counterpart. Moreover, loading the profile upfront catches invalid profiles during configuration loading rather than at container creation.

A notable behavior change is that if the file does not exist, it now falls back to the internal default instead of returning an error. This is consistent with `RuntimeConfig`, where the fallback logic was introduced in #6029.

@saschagrunert @bitoku Please take another look when you get a chance.

bitoku commented on pkg/config/config.go:
```diff
@@ -1310,15 +1321,20 @@ func (c *RuntimeConfig) Validate(systemContext *types.SystemContext, onExecution
 			}
 
-			logrus.Info("Specified profile does not exist on disk")
+			// Fallback to the internal default in order not to not break upgrade paths.
+			logrus.Info("Seccomp profile does not exist on disk, fallback to internal default profile")
```
Is this typo? The comment has duplicate "not". same for other places having this comment.

btw, @saschagrunert do you know why does it fall back to the default?
It seems introduced in this PR. https://github.com/cri-o/cri-o/pull/6029
I think it should raise an error, otherwise there's a risk of having a different seccomp profile without noticing it.

bitoku commented on pkg/config/config.go:
```diff
@@ -1999,6 +2025,30 @@ func (r *RuntimeHandler) validateRuntimeExecCPUAffinity() error {
 
+// validateRuntimeSeccompProfile tries to load the RuntimeHandler seccomp profile.
+func (r *RuntimeHandler) validateRuntimeSeccompProfile() error {
+	if r.SeccompProfile == "" {
+		r.seccompConfig = nil
+	} else {
```
nit: it can early-return and can reduce a nesting level.
```suggestion
	if r.SeccompProfile == "" {
		r.seccompConfig = nil
		return nil
	}
```

gavinkflam replied:
Done. However, our linters insist adding an empty line before return statement which isn't ideal in this case.

bitoku reviewed (commented):
seccomp `Setup()` should be closed to the receiver and shouldn't accept other `Config`.
Instead, how about returning the right seccomp in `Runtime.Seccomp()`?
I think this will make the code simpler.

Also this is nit, but it seems we can create a new function to validate seccomp. They look almost same.

bitoku reviewed (commented):
> To use RuntimeHandler.seccompConfig directly, we’d need to set up its notifierPath correctly. That requires APIConfig.Listen, which would mean passing the root Config struct deep into validateRuntimeSeccompProfile() - not ideal.

I don't think that using rootConfig is not ideal.
Cascading the value deep into `validateRuntimeSeccompProfile()` may look messy, but we can do just like this instead.
https://github.com/cri-o/cri-o/blob/51ce3dd659becb9d31c2752cf11c4411e3b18c22/pkg/config/config.go#L1063-L1065
Ideally, now that runtimeHandler has seccomp profile, everything should rely on runtimeHandler seccomp profile.

But this is more style suggestion, and if there's a reason not to do that, it's not a blocker of this PR.
We could also follow up if we find a better way.

/lgtm

bitoku commented on server/sandbox_run_linux.go:
```diff
@@ -866,6 +866,7 @@ func (s *Server) runPodSandbox(ctx context.Context, req *types.RunPodSandboxRequ
 			"",
 			nil,
 			nil,
+			nil,
```
Can we set it nil?

gavinkflam replied:
Good catch. I've implemented this for pods and added a test.

bitoku reviewed (commented):
/lgtm cancel
sorry there's one thing I missed.

gavinkflam commented:
@bitoku Please take another look.

I've added pod support, and implemented your suggestion of having `Runtime.Seccomp()` returning the right `seccomp.Config`.
